test/core/simd/meta/simd_integer_op.py

```diff
@@ -5,98 +5,184 @@
-        :param op: abs
-        :param p: a hex or decimal integer literal string
-        :lane_width: bit number of each lane in SIMD v128
-        :return:
+    def get_valid_value(value: int, lane: LaneValue, signed=True) -> int:
+        """Get the valid integer value in the scope of the specified lane size.
+
+        For a integer value, convert it to the valid value with the same bits
+        of the lane width. The value can be signed or unsigned, with the scope
+        of -0x80... to 0x7f... or 0 to 0xff...
+
+        :param value: the value of the integer
+        :param lane: the LaneValue instance of a lane in v128
+        :param signed: specify if the lane is interpreted as a signed or
+                        an unsigned number.
+        :return : the valid value in either signed or unsigned number
         """
-        if '0x' in p:
-            base = 16
-        else:
-            base = 10
-        v = int(p, base)
+        value &= lane.mask
+        if signed:
+            if value > lane.max:
+                return value - lane.mod
+            if value < lane.min:
+                return value + lane.mod
+        return value
+
+    def saturate(self, p1: int, p2: int, lane: LaneValue) -> int:
+        """Get the result of saturating arithmetic operation on 2 operands.
+        The operands can be both signed or unsigned. The following ops
+        are covered:
+        add_saturate_s, sub_saturate_s, add_saturate_u, sub_saturate_u,
+
+        Saturating arithmetic can make sure:
+        When the operation result is less than the minimum, return the minimum.
+        When the operation result is greater than the maximum, return the maximum.
+        For other operation results, simply return themselves.
+        :param p1: the integer operand 1
+        :param p2: the integer operand 2
+        :param lane: the LaneValue instance of a lane in v128
+        :return: the result of the saturating arithmetic operation
+        """
+        if self.op.endswith('saturate_s'):
+            if p1 > lane.max:
+                p1 -= lane.mod
+            if p2 > lane.max:
+                p2 -= lane.mod
+
+            if self.op.startswith('add'):
+                value = p1 + p2
+            if self.op.startswith('sub'):
+                value = p1 - p2
 
-        if op == 'abs':
-            result = IntegerSimpleOp.get_valid_value(v, lane_width)
+            if value > lane.max:
+                return lane.max
+            if value < lane.min:
+                return lane.min
+
+        if self.op.endswith('saturate_u'):
+            if p1 < 0:
+                p1 += lane.mod
+            if p2 < 0:
+                p2 += lane.mod
+            if self.op.startswith('add'):
+                value = p1 + p2
+            if self.op.startswith('sub'):
+                value = p1 - p2
+
+            if value > lane.mask:
+                return lane.mask
+            if value < 0:
+                return 0
+
+        return value
```

Comment:
If I understand correctly, this return should never be reached. Would it make more sense to make it an error?

Reply:
Actually, this `return` statement can be reached when the value is in a normal scope. The branches above can only handle the values out of boundary values.